server/app/services/question/ActiveAndDraftQuestions.java

```diff
@@ -24,39 +33,95 @@ public final class ActiveAndDraftQuestions {
           String, Pair<Optional<QuestionDefinition>, Optional<QuestionDefinition>>>
       versionedByName;
   private final ImmutableMap<String, DeletionStatus> deletionStatusByName;
+  private final Optional<ImmutableMap<String, ImmutableSet<ProgramDefinition>>>
+      referencingDraftProgramsByName;
+  private final ImmutableMap<String, ImmutableSet<ProgramDefinition>>
+      referencingActiveProgramsByName;
 
-  public ActiveAndDraftQuestions(Version active, Version draft) {
-    ImmutableMap.Builder<String, QuestionDefinition> activeToName = ImmutableMap.builder();
```

Comment:
Note to the reviewer: A good portion of the changes here were actually in using stream classes for better readability rather than having a bunch of ImmutableMap.Builders that were appended in a for-loop and built later on.